### ISSUE TITLE
Sparse-Dense Matrix Multiplication with CSR Format

### ISSUE THREAD

**Issue by CosimoRulli** [hfp/libxsmm#384]
Hello developers, 
I am interested in Sparse-Dense Matrix Multiplication using the CSR Format. I was wondering which is the algorithm  you followed to implement this routine. Is there any work in literature that describes it in detail? 

Just to give an example, I know that High Performance GEMM for dense matrix is usually based on the block pack algorithm described [here](https://www.cs.utexas.edu/~flame/pubs/GotoTOMS_revision.pdf). I am looking for something like that for sparse matrices.

**Comment by alheinecke** [comment]
Which sparse-dense matrix multiplication in libxsmm are you talking about? 

This: https://github.com/hfp/libxsmm/tree/master/samples/deeplearning/sparse_weight_mult 
or
This: https://github.com/hfp/libxsmm/tree/master/samples/spmdm

**Comment by CosimoRulli** [comment]
I would say that what I am looking for is [https://github.com/hfp/libxsmm/blob/master/samples/deeplearning/sparse_weight_mult/sparse_weight_A_mult.c](https://github.com/hfp/libxsmm/blob/master/samples/deeplearning/sparse_weight_mult/sparse_weight_A_mult.c). As I undertstood it, [this line](https://github.com/hfp/libxsmm/blob/f9b7ea189e3e7ba90521ebb5dde9795f2f1f28d1/samples/deeplearning/sparse_weight_mult/sparse_weight_A_mult.c#L140) generates the code which multiplies the sparse matrix A in the csr format with a dense matrix B ( l_b in the code).

**Comment by hfp** [comment]
Hello Cosimo,

you can run the example you are interested on and dump the generated code using `LIBXSMM_VERBOSE=-1`. For example:

```bash
LIBXSMM_VERBOSE=-1 ./sparse_weight_A_mult
[...]
LIBXSMM-JIT-DUMP(ptr:file) 0x2b4a25456000 : libxsmm_hsw_f32_nn_32x4x512_0_4_4_w16_a1_b1_p0_nnz1609.srsoa
[...]
```

As you can see the filename (`libxsmm_hsw_f32_nn_32x4x512_0_4_4_w16_a1_b1_p0_nnz1609.srsoa`) denotes the architecture the code was generated for as well as encoding other details. You can then proceed like:

```bash
objdump -D -b binary -m i386 -M x86-64 libxsmm_hsw_f32_nn_32x4x512_0_4_4_w16_a1_b1_p0_nnz1609.srsoa
```

This disassembles the code and shows the implementation to great detail. You can then also derive the approach taken: there is no data reordering or such but code that matches the sparsity structure of the input ("portrait"). Therefore, the generated kernel can only be used for the specific portrait however the non-zero values can be changed reusing the same kernel.

In summary, the approach taken is much different from your reference as it fully leverages what runtime code generation can do.

**Comment by alheinecke** [comment]
Not necessarily, the outer loop structure is very similar to libflame or BLIS: it's a dense tensor-contraction outer loop structure. The inner kernel is a sparse JIT'ed kernel hard-wiring the sparse structure of the corresponding block of the sparse matrix. I have created a couple of slides showing the idea. 
[2020_05_22_LIBXSMM_Sparse_JIT_public.pdf](https://github.com/hfp/libxsmm/files/4671016/2020_05_22_LIBXSMM_Sparse_JIT_public.pdf)

These paper might be helpful as well:
https://opt.dial3343.org/3343/edge_opt/-/blob/master/pub/posters/18_06_15_isc.pdf 
https://dial3343.org/pub/papers/19_03_26_isc_19.pdf

**Comment by CosimoRulli** [comment]
Thank you, the references really helped me out. I have another question: what is the difference between the two examples reported in this [answer](https://github.com/hfp/libxsmm/issues/384#issuecomment-632441369)? Is the first one for small matrices and the second for large ones?

**Comment by alheinecke** [comment]
No, the first one is ~4 years old and takes dense matrices as inputs and sparsifys them on the fly, it's not JITed and can handle various transposes and even activation sparsity. It's not that sensitive to matrix sizes and parallel. 

The second one (and the one I shared slides on) is for "small" sizes per core, but as the slides show on multi-core systems we can run large matmuls by running different "small" sizes on each core.

We can help you to pick the right kernel, if you can described you desired usage scenario.

**Comment by CosimoRulli** [comment]
I am experimenting different matrix multiplication methods to implement Fully Connected Forward Pass. I wanted to test the sparse-dense (A_sparse) and the dense-sparse routine (B_sparse) that you propose [here](https://github.com/hfp/libxsmm/tree/master/samples/deeplearning/sparse_weight_mult), both in single thread version. 

The A_sparse version gave me no problems. 
For what concerns the B_sparse version, when I compared its result with other multiplication routines, such as dnnl_sgemm from [OneDNN](https://github.com/oneapi-src/oneDNN) (aka MKL-DNN) or with classical three-loop based routine, I observed that the results differs, suggesting that one of the method is wrong.

To reproduce my observations, it is just needed to add this code

```
float * my_res= (float * )malloc(sizeof(float)*N*K);
for(i=0; i< N ;i++){
   for(j=0; j< K; j++){
      my_res[i*K + j] = 0;
      for(k=0; k< C; k++){
         my_res[i*K+j] += l_a[i*C+k]*l_b_de[j+k*K];
      }
   }
}
```
at the end of [this](https://github.com/hfp/libxsmm/blob/master/samples/deeplearning/sparse_weight_mult/sparse_weight_B_mult.c) example. 
If you execute the following lines  
```
for(i=0; i< N; i++){
     for(j=0; j<K; j++){
        diff1 = my_res[i*K+j] - l_c_asm_csr[i*K+j];
        diff2 = my_res[i*K+j] - l_c_gold[i*K+j];
        diff3 = my_res[i*K+j] - l_c_asm_csc[i*K+j];
        if (fabs(diff1) > toll) count1++;
        if (fabs(diff2) > toll) count2++;
        if (fabs(diff3) > toll) count3++;
     }
  }
```

you will note that all the three counters will be different from zero.

1. Line [109](https://github.com/hfp/libxsmm/blob/4ed82c58b529addee0731337d8dc467b0141667e/samples/deeplearning/sparse_weight_mult/sparse_weight_B_mult.c#L109) currently is:
`if ( LIBXSMM_VLA_ACCESS(2, l_p_b_de, l_i, l_j, C) != 0.0 )`
with l_j representing the row index and l_i the column index of a _C x K_ matrix that we would like to scan column-wise. So I think the correct code should be: 
`if ( LIBXSMM_VLA_ACCESS(2, l_p_b_de, l_j, l_i, K) != 0.0 )`

2. Line [126](https://github.com/hfp/libxsmm/blob/4ed82c58b529addee0731337d8dc467b0141667e/samples/deeplearning/sparse_weight_mult/sparse_weight_B_mult.c#L126) currently is :
` if ( LIBXSMM_VLA_ACCESS(2, l_p_b_de, l_i, l_j, C) != 0.0 )`
with this time l_j representing the column index and l_i the row index of the same _C x K_ matrix, that this time we want to scan row-wise. I would replace it with: 
`if ( LIBXSMM_VLA_ACCESS(2, l_p_b_de, l_j, l_i, K) != 0.0 ) `

3. Line [145](https://github.com/hfp/libxsmm/blob/4ed82c58b529addee0731337d8dc467b0141667e/samples/deeplearning/sparse_weight_mult/sparse_weight_B_mult.c#L145) currently is:
`* l_b_de[(l_j*C)+l_jj];`
with  l_j the column index and l_jj the row index of  l_b_de which is a  _C x K_ matrix, that we want to scan column-wise being the second matrix in a matrix multiplication. I think it should be: 
`* l_b_de[(l_jj*K) + l_j];`


With the previous changes, when n_b (the number blocks in which the A matrix is partitioned) is 1, the results are correct, but I could not manage to make this code work for any n_b. Are my corrections wrong? Do you have any advice?

**Comment by alheinecke** [comment]
l_b_de is a row-major stored matrix, means [k][n] in BLAS speech, or [K][C] in case of the example, https://github.com/hfp/libxsmm/blob/master/samples/deeplearning/sparse_weight_mult/sparse_weight_B_mult.c#L43 . 

Your tester reads B transposed (assuming B is [C][K]?) if I see that correctly:
` l_b_de[j+k*K]` 
as` j` loops over K and `k` over C.

So it's expected that your test fails as it compare to B^T multiplication, but all three variants in the examples multiply with B.

**Comment by CosimoRulli** [comment]
Yes, I was assuming B to be [C][K] since the operation we are performing is C=A*B, with A being [N][C] (as intialized [here](https://github.com/hfp/libxsmm/blob/a5b40e1ab209e53ec50b6609dd5eae267d224cf9/samples/deeplearning/sparse_weight_mult/sparse_weight_B_mult.c#L35)). So I thought that the shared dimension to be [C] to obtain a the final matrix C with shape [N][K].

**Comment by alheinecke** [comment]
Oh, https://github.com/hfp/libxsmm/blob/a5b40e1ab209e53ec50b6609dd5eae267d224cf9/samples/deeplearning/sparse_weight_mult/sparse_weight_B_mult.c#L35 it's just a malloc. It doesn't define memory layouts. The memory layouts are defined in the VLA declarations (https://github.com/hfp/libxsmm/blob/a5b40e1ab209e53ec50b6609dd5eae267d224cf9/samples/deeplearning/sparse_weight_mult/sparse_weight_B_mult.c#L43), and as B is sparse, there is no concept of row or col-major. We just re-used what we had and then the C loops for the reference code don't matter as it's all scalar accesses.

**Comment by CosimoRulli** [comment]
I agree with the fact the if B is sparse, row-major or col-major does not mind. But  there is still something I can not understand. 
If the operation we want to perform is C=A*B, and A is [N][C] (rearranged as as [NB][C][nb] tensor) and C is [N][K] (rearranged as a [NB][K][nb] tensor), the only shape that B can have is [C][K]. 

For example, the CSC format requires a _columns_ vector of length #columns +1. In fact, a K+1 vector is allocated to for the csc columns vector [here](https://github.com/hfp/libxsmm/blob/592d72f2902fa49fda51a4221e83f7f978936a9d/samples/deeplearning/sparse_weight_mult/sparse_weight_B_mult.c#L101). 

For the CSR, we need a _rows_ vector of length #rows +1, so a C+1 vector is allocated for csr [here](https://github.com/hfp/libxsmm/blob/592d72f2902fa49fda51a4221e83f7f978936a9d/samples/deeplearning/sparse_weight_mult/sparse_weight_B_mult.c#L118). 

If these assumptions are correct, my tester should be consistent. Or probably there is still something I did not understand correctly.

**Comment by alheinecke** [comment]
Sorry for the delay.

We are not using standard matrix layouts and yes the data is rearranged. However, the you pointed out before was in dense reference code and we don't need to reorder this one as this is scalar. So we do a fused transpose. 

Due to different accumulation orders you will get different errors for different versions (CSR vs. CSC).

**Comment by alheinecke** [comment]
closing for now.
